### PR TITLE
[Feature] Typesafe datatags

### PR DESCRIPTION
**Pull request by DerToaster98**
### Describe in detail what your pull request accomplishes
This PR adds typesafe datatags. Datatags work similar to craftfile properties in that they need to be registered before using them. Their value can be accessed via methods on the craft directly.

### Checklist
- [ ] Unit tests
- [ ] Proper internationalization
- [x] Tested

## TODO list:
- ~~API classes~~
- ~~Access methods~~
- Calling update methods of the value in the respective spots
  - onRotate
  - onSwitchWorld
  - onRelease
  - onAssembly
  - onTranslate

### Related PRs:
- #611

### CODE REVIEW

**Review by TylerS1066 — CHANGES_REQUESTED** [review]
Looking at this PR, I think there's a difference in what we view craft data tags to be.  I view them as being type-safe data associated with a craft by a namespaced key.

The first and foremost of this would be the contacts system, I envision `ContactsManager` would have the following line:
`public static final NamespacedKey CONTACTS = new NamespacedKey("movecraft", "contacts");`
Then anywhere you need to get the contacts from a craft you would run the following:
`craft.getObjectTag(ContactsManager.CONTACTS)` to return an `Object`.  We'll have to decide what types we want explicit type safety on (since it's a lot of duplicate code for each), but here's my initial list that should cover most data in crafts:
- boolean
- double
- long
- String
- Object (like craft properties, type safety will be the responsibility of the user)

File: api/src/main/java/net/countercraft/movecraft/craft/datatag/CraftDataTagContainer.java
Comment (DerToaster98) [reply api/src/main/java/net/countercraft/movecraft/craft/datatag/CraftDataTagContainer.java:22]: Well i do that to avoid stuff like this:
CraftDataTagKey<String> A = REGISTER("bob", ...);
CraftDataTagKey<Boolean> B = REGISTER("bob", ...)

Now, when i want to retrieve A using .get() it will throw some casting exception. ALso the way this works, the registered tags basically act like a default value creation supplier for the map.

File: Movecraft/src/main/java/net/countercraft/movecraft/craft/BaseCraft.java
Comment (DerToaster98) [reply Movecraft/src/main/java/net/countercraft/movecraft/craft/BaseCraft.java:566]: Ye, we can remove that, .get() and .set() is enough

File: api/src/main/java/net/countercraft/movecraft/craft/datatag/ICraftDataTag.java
Comment (DerToaster98) [reply api/src/main/java/net/countercraft/movecraft/craft/datatag/ICraftDataTag.java:11]: Idea was to allow Object datatags to directly react to these events, but yes, this can be replicated with the event API pretty quickly

**Comment by DerToaster98** [comment]
> Looking at this PR, I think there's a difference in what we view craft data tags to be. I view them as being type-safe data associated with a craft by a namespaced key.
> 
> The first and foremost of this would be the contacts system, I envision `ContactsManager` would have the following line: `public static final NamespacedKey CONTACTS = new NamespacedKey("movecraft", "contacts");` Then anywhere you need to get the contacts from a craft you would run the following: `craft.getObjectTag(ContactsManager.CONTACTS)` to return an `Object`. We'll have to decide what types we want explicit type safety on (since it's a lot of duplicate code for each), but here's my initial list that should cover most data in crafts:
> 
>     * boolean
> 
>     * double
> 
>     * long
> 
>     * String
> 
>     * Object (like craft properties, type safety will be the responsibility of the user)

Well what i expect is this: 

`CraftDataTagKey<Contacts> CONTACTS_KEY = register("contacts", Contacts::new) `


If i want the contacts, i can just do 
`Contacts contacts = Craft.GetDataTag(CONTACTS_KEY)`

That way, no cast is necessary and i can directly access it.
I don't see dedicated types for bool, double, etc as necessary, for those one can just use the wrapper classes. Also havign dedicated classes for that in my opinion just adds bloat code. What i can live with is additional methods for the primitives and one getObject method that works like my example above (aka the current implementation). Otherwise i agree with your definition of craft data tags, for me it is just additional data that a craft drags along for itself, in addition it would be neat if that data can directly react to some events like the craft being piloted or moved.

**Comment by TylerS1066** [comment]
That's an interesting approach and I like the implicit type safety built in.  Is it possible we could change the supplier to a `Function<Craft, ...>` that way you can calculate the value based on the craft?

**Comment by DerToaster98** [comment]
> That's an interesting approach and I like the implicit type safety built in. Is it possible we could change the supplier to a `Function<Craft, ...>` that way you can calculate the value based on the craft?

Yes, sure, that's a good idea!

File: Movecraft/src/main/java/net/countercraft/movecraft/craft/BaseCraft.java
Comment (DerToaster98) [reply Movecraft/src/main/java/net/countercraft/movecraft/craft/BaseCraft.java:561]: Actually, java disallows private and protected modifiers in interfaces. Since that method is used by the interface default implementations of getDataTag() and setDataTag(), we also can't remove it

Comment (TylerS1066) [reply Movecraft/src/main/java/net/countercraft/movecraft/craft/BaseCraft.java:561]: Is there a reason we need to expose the `CraftDataTagContainer` at the craft level?  I think we could just leave it totally internal to `BaseCraft`.

File: Movecraft/src/main/java/net/countercraft/movecraft/craft/CruiseOnPilotSubCraft.java
Comment (TylerS1066) [inline Movecraft/src/main/java/net/countercraft/movecraft/craft/CruiseOnPilotSubCraft.java:26]: I was going to clean up this diff myself, but it looks like "allow edits from maintainers" wasn't selected when this PR was opened.

File: api/src/main/java/net/countercraft/movecraft/craft/datatag/CraftDataTagContainer.java
Comment (TylerS1066) [inline api/src/main/java/net/countercraft/movecraft/craft/datatag/CraftDataTagContainer.java:18]: Rather than be `tryRegisterTagKey` we could rename this to be `registerTagKey` and annotate it with `throws IllegalArgumentException`.

**Comment by TylerS1066** [comment]
I've tried out this system to implement contacts and I much prefer it over the map it used to contain!  Once this PR is cleaned up and merged I'll rebase [these changes](https://github.com/APDevTeam/Movecraft/tree/contacts-data-tag) and merge them as well.

File: api/src/main/java/net/countercraft/movecraft/craft/datatag/CraftDataTagContainer.java
Comment (DerToaster98) [reply api/src/main/java/net/countercraft/movecraft/craft/datatag/CraftDataTagContainer.java:18]: Ye, that would be cleaner, also doesn't lead to unexpected exceptions

File: Movecraft/src/main/java/net/countercraft/movecraft/craft/CruiseOnPilotSubCraft.java
Comment (DerToaster98) [reply Movecraft/src/main/java/net/countercraft/movecraft/craft/CruiseOnPilotSubCraft.java:26]: I'll see what i can do

Comment (DerToaster98) [reply Movecraft/src/main/java/net/countercraft/movecraft/craft/CruiseOnPilotSubCraft.java:26]: CraftDataTagContainer being at craft level is cause the set and get methods are implemented using default methods in the interface, which requires access to the container in a way. And since interfaces can't have fields we need some method for it and the craft needs to implement it to give access to the field.

**Comment by DerToaster98** [comment]
> I've tried out this system to implement contacts and I much prefer it over the map it used to contain! Once this PR is cleaned up and merged I'll rebase [these changes](https://github.com/APDevTeam/Movecraft/tree/contacts-data-tag) and merge them as well.

I was also planning to use this for various changes i want to make to movecraft combat (mainly a overall rework of directors), glad you like it.

**Comment by DerToaster98** [comment]
Removed the unnecessary interface, that interface is only really useful if we want to directly support datatypes that can directly react to events

**Review by TylerS1066 — CHANGES_REQUESTED** [review]
Only thing left is to clear up the diff with the following files:
- `Movecraft/src/main/java/net/countercraft/movecraft/craft/CruiseOnPilotSubCraft.java`
- `Movecraft/src/main/java/net/countercraft/movecraft/craft/SinkingCraftImpl.java`
- `Movecraft/src/main/java/net/countercraft/movecraft/craft/SubCraftImpl.java`
- `Movecraft/src/main/java/net/countercraft/movecraft/craft/SubcraftRotateCraft.java`

**Comment by DerToaster98** [comment]
Done

**Review by TylerS1066 — APPROVED** [review]

**Pull request merged**
